runner/src/lib.rs

```diff
@@ -7,4 +7,9 @@ pub use runtime::SandboxRuntime;
 #[cfg(not(test))] // Work around for rust-lang/rust#62127
 pub use runner_macros::main;
 
+// Used for generated code, Not a public API
+#[doc(hidden)]
+#[path = "private/mod.rs"]
+pub mod __private;
```

Comment:
I have 0.9 strong feeling that macros are a wrong approach here, but, as I don't have domain knowledge, my confidence level here is only about 0.5. Let me try to explain my reasoning.

>Was pretty much just going off of what tokio had and having the nicety of setting up the whole runtime for the dev.

I don't think this is a good reason to use macros, as context needs to be taken into account. Tokio is, in a sense, a whole world. It is an alternative and replacement for much of the standard library which provides a different way to do IO and concurrency. As such, tokio has much bigger reason to use macros -- part of "io replacement" is replacing thread-based main with epoll-loop based main. It also has significantly less costs for using macros -- everyone knows and uses tokio, so everyone knows the macros already. 

Here, we are not building a language runtime -- we are building a client for an RPC service. Good examples to learn from would be other popular crate which provides rust bindings to web services. I am not familiar with this space, but one thing that comes to mind are AWS SDKs in particular, aws lambda feels pretty close to what we want here. Here's invoke lambda example:

https://github.com/awslabs/aws-sdk-rust/blob/acfa8b17206870802e0074a0869147cada5550b5/sdk/examples/lambda/src/bin/invoke-function.rs#L33-L63

It look as I expect: we do have `#[tokio::main]`, as we are async, but other than that, that's just boring pedestrian code -- we build a config struct, using this config we initialize the `client` object. All the API is then available as methods on this `client` object. That's roughly the shape I'd expect to see in the runner. 

<hr/>

That was the general philosophy for why copying tokio makes sense only if you are a similarly big-city-named crate. Some specific reasons to avoid macros:

* They are very opaque for developers. As a new user of the crate, I have no idea what `#[runner::test(sandbox)]` does, and I don't have a way to figure that out. I can press go-to definition, but its impossible to decipher [this proc macro impl](https://github.com/near/workspaces-rs/blob/9dc13d566ed40ac38dc4e9251d5ca27847ecf8ad/runner-macros/src/entry.rs#L11-L127) into what actually happens there. 
* They are tough for IDE support. Basically, the IDE has the choice between ignoring the macro completely and analysing the wrong code, and expanding the macro and forgoing error resilience, as proc-macros break for partiall-incomplete code
* Longer compile times
* Poor composability -- if I want to abstract over what the macro is doing, I'd have to write yet another macro. I can't use traits and generics 

To generalize, proc-macros are a multiplier for accidental complexity for people *using* the crates. As such, they are mostly justified when there's a physical reason to use them. In case of tokio, that's the desire to replace threads with epoll. With near-sdk-rs, that's the necessity to emit a bunch of `extern "C"` functions in a type-safe way, in case of [sqlx](https://github.com/launchbadge/sqlx), that's the desire to talk with database at compile times. 

If the goal is to just reduce the amount of boilerplate, almost always a better solution is to design API to minimize boilerplate -- add some helper shortcut functions which make common operations short.

Reply:
Definitely all good points, I think the primary benefit to the current setup is that you don't require being aware and pulling in tokio as well as calling some function within the test/main to initialize the sandbox. 

I definitely was under the impression that this functionality would have to be under a feature like `macros` is for tokio before being released, but I think you are correct in that we have placed too much of a priority on this rather than the core interface without macros.